core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -277,6 +277,13 @@ public class SqlToRelConverter {
   private final Map<CorrelationId, DeferredLookup> mapCorrelToDeferred =
       new HashMap<>();
 
+  /**
+   * The scope and identifier to the correlationId to deduplicate the creation of the
+   * correlationId.
+   */
+  private final Map<Pair<SqlValidatorScope, String>, CorrelationId> correlationIdMap =
```

Comment:
This map will rely on the object identifier of the scope, in the test case I posted, It can work well. But I'm not sure whether it's a good enough solution to reply on this object identifier here. 

Looking forward to some suggestion, thanks.